src/OrleansRuntime/Streams/QueueBalancer/LeaseBasedQueueBalancer.cs

```diff
@@ -0,0 +1,345 @@
+    /// as much as possible. 
+    /// </summary>
+    public class LeaseBasedQueueBalancer : QueueBalancerBaseClass, ISiloStatusListener, IStreamQueueBalancer, IDisposable
+    {
+        private class AcquiredQueue 
+        {
+            public QueueId QueueId { get; set; }
+            public AcquiredLease AcquiredLease { get; set; }
+            public AcquiredQueue(QueueId queueId, AcquiredLease lease)
+            {
+                this.QueueId = queueId;
+                this.AcquiredLease = lease;
+            }
+        }
+        private ILeaseProvider leaseProvider;
+        private IDeploymentConfiguration deploymentConfig;
+        private ISiloStatusOracle siloStatusOracle;
+        private ReadOnlyCollection<QueueId> allQueues;
+        private List<AcquiredQueue> myQueues;
+        private TimeSpan siloMaturityPeriod;
+        private bool isStarting;
+        private ConcurrentDictionary<SiloAddress, bool> immatureSilos;
+        private TimeSpan leaseLength = TimeSpan.FromMinutes(1);
+        private AsyncTaskSafeTimer renewLeaseTimer;
+        private AsyncTaskSafeTimer tryAcquireMaximumLeaseTimer;
+        private IResourceSelector<QueueId> queueSelector;
+        private int minimumResponsibilty;
+        private int maximumRespobsibility;
+        private Logger logger;
+
+        public LeaseBasedQueueBalancer(ILeaseProvider leaseProvider, ISiloStatusOracle siloStatusOracle, IDeploymentConfiguration deploymentConfig, Factory<string, Logger> loggerFac)
+            : base()
+        {
+            this.leaseProvider = leaseProvider;
+            this.deploymentConfig = deploymentConfig;
+            this.siloStatusOracle = siloStatusOracle;
+            this.myQueues = new List<AcquiredQueue>();
+            this.immatureSilos = new ConcurrentDictionary<SiloAddress, bool>();
+            this.isStarting = true;
+            this.logger = loggerFac(this.GetType().Name);
+            // register for notification of changes to silo status for any silo in the cluster
+            this.siloStatusOracle.SubscribeToSiloStatusEvents(this);
+            // record all already active silos as already mature. 
+            // Even if they are not yet, they will be mature by the time I mature myself (after I become !isStarting).
+            foreach (var silo in siloStatusOracle.GetApproximateSiloStatuses(true).Keys.Where(s => !s.Equals(siloStatusOracle.SiloAddress)))
+            {
+                immatureSilos[silo] = false;     // record as mature
+            }
+        }
+
+        public override Task Initialize(string strProviderName, IStreamQueueMapper queueMapper, TimeSpan siloMaturityPeriod)
+        {
+            if (queueMapper == null)
+            {
+                throw new ArgumentNullException("queueMapper");
+            }
+            this.allQueues = new ReadOnlyCollection<QueueId>(queueMapper.GetAllQueues().ToList());
```

Comment:
There's no ReadOnlyList :( 

I will make the thing in RoundRobinSelector to be ReadOnlyCollection to avoid extra allocation,